localstack/services/s3/cors.py

```diff
@@ -0,0 +1,250 @@
+import re
+from typing import Optional, Tuple
+
+from werkzeug.datastructures import Headers
+
+from localstack import config
+from localstack.aws.api import RequestContext
+from localstack.aws.api.s3 import AccessForbidden, BadRequest, CORSRule, CORSRules
+from localstack.aws.chain import Handler, HandlerChain
+
+# TODO: refactor those to expose the needed methods
+from localstack.aws.handlers.cors import CorsEnforcer, CorsResponseEnricher
+from localstack.aws.protocol.op_router import RestServiceOperationRouter
+from localstack.aws.protocol.serializer import gen_amzn_requestid_long
+from localstack.aws.protocol.service_router import get_service_catalog
+from localstack.constants import S3_VIRTUAL_HOSTNAME
+from localstack.http import Request, Response
+from localstack.services.s3.models import BucketCorsIndex
+from localstack.services.s3.utils import S3_VIRTUAL_HOSTNAME_REGEX
+
+_s3_virtual_host_regex = re.compile(S3_VIRTUAL_HOSTNAME_REGEX)
+FAKE_HOST_ID = "9Gjjt1m+cjU4OPvX9O9/8RuvnG41MRb/18Oux2o5H5MY7ISNTlXN+Dz9IG62/ILVxhAGI0qyPfg="
+
+# TODO: refactor those to expose the needed methods maybe in another way that both can import
+add_default_headers = CorsResponseEnricher.add_cors_headers
+is_origin_allowed_default = CorsEnforcer.is_cors_origin_allowed
+
+
+class S3CorsHandler(Handler):
+
+    bucket_cors_index: BucketCorsIndex
+
+    def __init__(self):
+        self.bucket_cors_index = BucketCorsIndex()
+        self._service = get_service_catalog().get("s3")
+        self._s3_op_router = RestServiceOperationRouter(self._service)
+
+    def __call__(self, chain: HandlerChain, context: RequestContext, response: Response):
+        self.handle_cors(chain, context, response)
+
+    def pre_parse_s3_request(self, context: RequestContext) -> Tuple[bool, Optional[str]]:
+        is_s3: bool
+        bucket_name: str
+
+        path = context.request.path
+        host = context.request.host
+
+        # first, try to figure out best-effort whether the request is an s3 request
+        if host.startswith(S3_VIRTUAL_HOSTNAME):
+            is_s3 = True
+            bucket_name = path.split("/")[1]
+        # try to extract the bucket from the hostname (the "in" check is a minor optimization
+        elif ".s3" in host and (match := _s3_virtual_host_regex.match(host)):
+            is_s3 = True
+            bucket_name = match.group(3)
+        # otherwise we're not sure, and whether it's s3 depends on whether the bucket exists. check later
+        else:
+            is_s3 = False
+            bucket_name = path.split("/")[1]
+
+        existing_buckets = self.bucket_cors_index.buckets
+        if bucket_name not in existing_buckets:
```

Comment:
I'm not entirely sure if the cache handling here is correct.
- The cache is built the first time it is accessed (and everytime it is accessed after it has been invalidated).
- The cache is _not_ invalidated on the creation of a bucket.

What happens if:
1. A bucket and CORS rules for the bucket are created.
2. The bucket is requested (the cache is filled and not invalidated).
3. A new bucket is created (the cache is not invalidated).
4. The new bucket is requested.

Then this check here would not be correct, right? It would basically say "there is no bucket" even if there is one?

Reply:
I think the cache is invalidated on bucket creation (line 225 in `provider.py`). But that would have been problematic yes

Reply:
Ah, thanks! I missed the `self._cors_handler.invalidate_cache()` in `create_bucket`.